### PR TITLE
Add tests for RegExp set notation

### PR DESCRIPTION
**Pull request by mathiasbynens**
Set notation and extended character class functionality is part of the RegExp `v` flag proposal: https://github.com/tc39/proposal-regexp-set-notation

Issue: https://github.com/tc39/test262/issues/3496, https://github.com/tc39/proposal-regexp-set-notation/issues/52

Test generator code: https://github.com/mathiasbynens/unicode-property-escapes-tests/pull/11/files

### CODE REVIEW

**Comment by mathiasbynens** [comment]
I wasn’t sure about the directory structure. The current PR uses `test/built-ins/RegExp/unicodeSets/generated/*.js`. With that, we get the following:

- Tests that are specific to the `v` flag (now and forever) live under `test/built-ins/RegExp/unicodeSets/*`. (`unicodeSets` is the name of the flag getter on `RegExp.prototype`, matching the precedent in that directory.)
- All tests that are generated live under a `generated` subdirectory, to clearly distinguish them from manually maintained tests.
- `test/built-ins/RegExp/property-escapes/*.js` contains all `\p{…}` tests — both for character properties with the `u` flag, and for properties of strings with the `v` flag. Rationale: once the `v` flag proposal is Stage 4, we want to bring properties of strings to `u` RegExps as well: https://github.com/tc39/proposal-regexp-set-notation/issues/49#issuecomment-988956561 cc @bakkot In other words, even though properties of strings are _currently_ only available for `v` RegExps, they are planned to be made available for `u` RegExps as well, so it doesn’t make sense to put them under a flag-specific directory.

I’m happy to change any of this — just wanted to share my current thinking.

**Comment by ptomato** [comment]
@mathiasbynens Thanks for putting all this together. Since you're the expert on the proposal, do you have a suggestion of what kind of review you're looking for at this stage?

**Comment by mathiasbynens** [comment]
@ptomato I might be involved with the proposal, but I make mistakes like everyone else :) I’d love to get a review, _especially_ from folks who are not currently familiar with the proposal. For each test, check the `regExp` property, and double-check that our proposal’s grammar and semantics allow this pattern. Then, check it against some of the `matchStrings` and `nonMatchStrings` (not exhaustively), and see if anything looks off.

**Comment by rwaldron** [comment]
@mathiasbynens would mind @ mention/summoning folks that you think would be ideal reviewers? Thanks!

**Comment by mathiasbynens** [comment]
> @mathiasbynens would mind @ mention/summoning folks that you think would be ideal reviewers? Thanks!

Literally anyone

**Comment by ptomato** [comment]
Sounds good. We talked about this in the maintainers group today and several of us will try to take a non-exhaustive look at this in the coming weeks.

**Comment by mathiasbynens** [comment]
@ptomato Amazing — thank you! Looking forward to hearing your feedback.

**Comment by mathiasbynens** [comment]
@ptomato Any updates for us? How can we get this moving forward? V8 is planning to implement the proposal in Q3 and it’d be helpful to have these Test262 tests landed by then.

**Comment by ptomato** [comment]
I expect to be able to devote some attention to this starting next week.

In the meantime it occurred to me that it might be helpful to involve those who are implementing this proposal in other engines for reviews — do you know of any?

**Comment by mathiasbynens** [comment]
> In the meantime it occurred to me that it might be helpful to involve those who are implementing this proposal in other engines for reviews — do you know of any?

@pthier @iainireland @msaboff Would you be interested in taking a look?

Disclaimer: At least for V8, it’s more common for implementers to review the tests while working on the implementation (instead of beforehand). So merging the tests after @ptomato’s review, and then updating them later as needed, could be another option.

**Review by ptomato — APPROVED** [review]
I went through this over yesterday and today. I didn't check everything exhaustively but I looked at each file. From my point of view I think it can be merged. I don't know if anyone else wanted to look at it?

My compliments for the choice of character classes to be combined with each other. I thought they were very well chosen to exercise the three combination operators.

I found one weirdness (harmless, though) which might indicate a bug in the test generator scripts.

Thanks for putting all this together.

File: test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character-class-escape.js
Comment (mathiasbynens) [reply test/built-ins/RegExp/unicodeSets/generated/character-class-difference-character-class-escape.js]: Done.

**Comment by mathiasbynens** [comment]
> I don't know if anyone else wanted to look at it?

Given that this has been open for a while and no one else stepped up, I somewhat doubt it. Perhaps we could go ahead and merge this as-is? I’d be happy to address any follow-up issues separately.

**Review by ptomato — APPROVED** [review]
Re-approving with bug fix

**Pull request merged**
